src/Elastic.Apm/Helpers/DbConnectionStringParser.cs

```diff
@@ -0,0 +1,234 @@
+			new Dictionary<string, Action<string, Destination>>(StringComparer.OrdinalIgnoreCase)
+			{
+				{ "Server" , ParseServerValue },
+				{ "Data Source" , ParseServerValue },
+				{ "Host" , ParseServerValue },
+				{ "Hostname" , ParseServerValue },
+				{ "Network Address" , ParseServerValue },
+				{ "dbq" , ParseServerValue },
+				{ "Port" , ParsePortValue }
+			};
+
+		private const char KeyValuePairsSeparator = ';';
+		private const char KeyValueSeparator = '=';
+		private const char ServerNameDbInstanceSeparator = '\\';
+		private const string SqlServerLocalDbPrefix = "(LocalDB)";
+		private const string SqlServerExpressUserInstancePrefix = @".\";
+		private const string SqlAzurePrefix = @"tcp:";
+		private const string OracleXeClientSuffix = @"/XE";
+		private static readonly IEnumerable<string> DiscardablePrefixes = new List<string>
+		{
+			SqlAzurePrefix
+		};
+		private static readonly IEnumerable<string> DiscardableSuffixes = new List<string>
+		{
+			OracleXeClientSuffix
+		};
+
+		/// <returns><c>Destination</c> if successful and <c>null</c> otherwise</returns>
+		private Destination ParseConnectionString(string dbConnectionString)
+		{
+			Destination result = null;
+			foreach (var keyValueString in dbConnectionString.Split(KeyValuePairsSeparator))
+			{
+				if (string.IsNullOrWhiteSpace(keyValueString)) continue;
+
+				var keyValueArray = keyValueString.Split(KeyValueSeparator);
+				if (keyValueArray.Length != 2)
+				{
+					_logger.Trace()?.Log("Encountered invalid key-value pair - skipping it."
+						+ " keyValueString: `{KeyValueString}'. keyValueArray: {KeyValueArray}. dbConnectionString: {DbConnectionString}."
+						, keyValueString, keyValueArray, dbConnectionString);
+					continue;
+				}
+
+				if (! KeyToPropertySetter.TryGetValue(keyValueArray[0].Trim(), out var propSetter)) continue;
+
+				if (result == null) result = new Destination();
+
+				try
+				{
+					propSetter(keyValueArray[1].Trim(), result);
+				}
+				catch (Exception ex)
+				{
+					_logger.Trace()?.LogException(ex, "Encountered invalid value for a known key"
+						+ " - considering the whole connection string as invalid and returning null."
+						+ " keyValueString: `{KeyValueString}'. keyValueArray: {KeyValueArray}. dbConnectionString: {DbConnectionString}."
+						, keyValueString, keyValueArray, dbConnectionString);
+					return null;
+				}
+			}
+
+			return result;
+		}
+
+		private static void ParseServerValue(string valueToParseArg, Destination destination)
+		{
+			var valueToParse = TrimDiscardable(valueToParseArg);
+
+			if (valueToParse.StartsWith(SqlServerLocalDbPrefix, StringComparison.OrdinalIgnoreCase)
+				|| valueToParse.StartsWith(SqlServerExpressUserInstancePrefix, StringComparison.OrdinalIgnoreCase))
+			{
+				destination.Address = "localhost";
+				return;
+			}
+
+			var dbInstanceSeparatorIndex = valueToParse.IndexOf(ServerNameDbInstanceSeparator);
```

Comment:
I assume you refer to the test failing because `ELASTIC_APM_TESTS_MS_SQL_HOST` contained DB instance name in addition to the host name - I've added a comment above.